### PR TITLE
Only use typeid() in std.h when it is available

### PR DESCRIPTION
**Pull request by danakj**
This allows the std.h formatters to work when RTTI is disabed.

For issue #3464

### CODE REVIEW

**Comment by danakj** [comment]
https://github.com/chromium/subspace/actions/runs/5127988704/ demonstrates this PR at work with RTTI disabled

File: include/fmt/core.h [hunk 1]
```diff
@@ -1786,7 +1804,7 @@ class format_arg_store
         basic_format_args<Context>(*this),
 #endif
         data_{detail::make_arg<is_packed, Context>(args)...} {
-    if (num_named_args != 0)
+    if constexpr (num_named_args != 0)
```
Comment (mwinterb) [inline include/fmt/core.h:1807]: Using `const_check` is generally preferred for this warning avoidance.
https://github.com/fmtlib/fmt/blob/bd393456ed94a5d64bb6078aa3fa477ceeb621d9/include/fmt/core.h#L319-L320

Comment (danakj) [reply include/fmt/core.h:1807]: Thank you, done.

File: include/fmt/core.h [hunk 2]
```diff
@@ -138,6 +138,24 @@
...
+#  elif defined(__INTEL_RTTI__)  // Intel compiler.
+#    define FMT_USE_TYPEID 1
+#  elif defined(__GXX_RTTI)  // G++.
+#    define FMT_USE_TYPEID 1
+#  elif FMT_HAS_FEATURE(cxx_rtti)  // Clang.
+#    define FMT_USE_TYPEID 1
+#  else
+#    define FMT_USE_TYPEID 0
+#  endif
+#endif
```
Comment (vitaut) [inline include/fmt/core.h:157]: Please move this to fmt/std.h where it is used.

File: include/fmt/core.h [hunk 3]
```diff
@@ -1786,7 +1804,7 @@ class format_arg_store
         basic_format_args<Context>(*this),
 #endif
         data_{detail::make_arg<is_packed, Context>(args)...} {
-    if (num_named_args != 0)
+    if (detail::const_check(num_named_args != 0))
```
Comment (vitaut) [inline include/fmt/core.h:1807]: Unrelated change, please revert.

File: include/fmt/std.h [hunk 4]
```diff
@@ -274,7 +274,9 @@ struct formatter<
     if (it == end || *it == '}') return it;
     if (*it == 't') {
       ++it;
+#if FMT_USE_TYPEID
       with_typename_ = true;
+#endif
```
Comment (vitaut) [inline include/fmt/std.h:279]:
```
with_typename_ = FMT_USE_TYPEID != 0;
```

Comment (danakj) [reply include/fmt/std.h:279]: Done.

File: include/fmt/core.h (hunk 2 above)
Comment (danakj) [reply include/fmt/core.h]: Done.

File: include/fmt/core.h (hunk 3 above)
Comment (vitaut) [reply include/fmt/core.h:1789]: Separate commit is fine.

**Review by vitaut — APPROVED** [review]

**Comment by danakj** [comment]
Thank you as well :)
